pkg/controllers/static/provisioning/controller.go

```diff
@@ -73,6 +73,13 @@ func (c *Controller) Reconcile(ctx context.Context, np *v1.NodePool) (reconcile.
 		return reconcile.Result{}, nil
 	}
 
+	// We need to wait until our representation of cluster is populated accurately with what we have in api-server or else
+	// we would end up over provisioning due to misrepresentation of NodePoolState in our cluster state.
+	// This usually happens when there are controller crashes
+	if !c.cluster.Synced(ctx) {
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
```

Comment:
One of the advantages of our current approach is that we don't need to await cluster sync every time we do a provisioning action. I think we'd like to maintain that if possible. I think the most straightforward approach to fix this issue and achieve this property would be to introduce a separate `Synced` method on NodePoolState and that is a "sticky" wrapper around `cluster.Synced()`. Once it returns true once, it will always return true without rechecking.

Reply:
I think we can probably do better than this, but I can see us punting that for the longer term effort to break up cluster state.

Reply:
Cluster State has `HasSynced` to check if we have synced atleast once, gonna use that. 
HasSynced will be true and Synced will be false in the case where we have nodeclaims without providerID, we wait until NodeClaims resolve providerID values to mark as Synced, in this case we dont really care as we mark the NodeClaim as Active in NodePool state regardless of weather we have resolved NodeClaim providerIDs